httpclient5/src/main/java/org/apache/hc/client5/http/config/ConnectionMgmtConfig.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.client5.http.config;
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+import org.apache.hc.core5.http.HttpHost;
+import org.apache.hc.core5.util.Timeout;
+
+/**
+ *  Immutable class encapsulating connection initialization and management settings.
+ *
+ * @since 5.2
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class ConnectionMgmtConfig implements Cloneable {
+
+    private static final Timeout DEFAULT_CONNECTION_REQUEST_TIMEOUT = Timeout.ofMinutes(3);
+    private static final Timeout DEFAULT_CONNECT_TIMEOUT = Timeout.ofMinutes(3);
+
+    public static final ConnectionMgmtConfig DEFAULT = new Builder().build();
+
+    private final HttpHost proxy;
+    private final Timeout socketTimeout;
+    private final Timeout connectionRequestTimeout;
+    private final Timeout connectTimeout;
+
+    /**
+     * Intended for CDI compatibility
+    */
+    protected ConnectionMgmtConfig() {
+        this(null, null, DEFAULT_CONNECTION_REQUEST_TIMEOUT, DEFAULT_CONNECT_TIMEOUT);
+    }
+
+    ConnectionMgmtConfig(
+            final HttpHost proxy,
+            final Timeout socketTimeout,
+            final Timeout connectionRequestTimeout,
+            final Timeout connectTimeout) {
+        super();
+        this.proxy = proxy;
+        this.socketTimeout = socketTimeout;
+        this.connectionRequestTimeout = connectionRequestTimeout;
+        this.connectTimeout = connectTimeout;
```

Comment:
My understanding is that there are two sets of configuration options:
1. Options which apply to the connection pool as a whole
2. Options that may differ based on individual requests

I think the `connectTimeout` should generally apply to an entire pool of connections (within a route anyhow), while the `connectionRequestTimeout` is more likely to be configured on a per-request basis. Although if pooling is disabled, that may not be the case, but in the modern era of http/1.1 and greater it's probably reasonable to assume pooling is used.

It's not clear to me that all of these options should be configurable per-request, while others should. So I suppose my questions are:
Should this be split into two separate configurations for per-pool/route configuration, and per-request configuration? 
Should the division between `ConnectionMgmtConfig` and `RequestConfig` be different?

Reply:
> Options which apply to the connection pool as a whole

It seems that the current convention is to specify those options directly on the connection manager, e.g. connection TTL, idle connection validation, connection pool queueing and concurrency policies, etc. So if we wanted to make `connectTimeout` specify behavior for the pool as a whole without the ability to overload `connectTimeout` on a per-request basis, I assume we would add that option to `PoolingHttpClientConnectionManagerBuilder` and similar classes.

Reply:
I suppor the config on the pool itself. Using Commons Pool or Tomcat JDBC Pool I specify how long it should block until a get a valid option from. This is an immutable config on the pool and not on the request. This should be completely opaque to the client code.

Reply:
@carterkozak @rschmitt @michael-o All right. I will see if I can move `ConnectionMgmtConfig` to connection managers, though I am sure it is a matter of time till someone pops up on the mailing list and demands those settings to be made available on a per request basis.

Reply:
@ok2c I think the challenge here is that the connect timeout should actually be or additionally be on a per-route basis because the conn manager handles connections to different servers w/o different attributes in contrast to a database conn pool which in most cases connects to a single host. WDYT?